pkg/reconciler/pipelinerun/pipelinerun.go

```diff
@@ -1183,116 +1198,167 @@ func (c *Reconciler) updatePipelineRunStatusFromInformer(ctx context.Context, pr
 		logger.Errorf("could not list TaskRuns %#v", err)
 		return err
 	}
-	updatePipelineRunStatusFromTaskRuns(logger, pr, taskRuns)
-
 	runs, err := c.runLister.Runs(pr.Namespace).List(k8slabels.SelectorFromSet(pipelineRunLabels))
 	if err != nil {
 		logger.Errorf("could not list Runs %#v", err)
 		return err
 	}
-	updatePipelineRunStatusFromRuns(logger, pr, runs)
+
+	updatePipelineRunStatusFromSlices(ctx, logger, pr, taskRuns, runs)
 
 	return nil
 }
 
-func updatePipelineRunStatusFromTaskRuns(logger *zap.SugaredLogger, pr *v1beta1.PipelineRun, trs []*v1beta1.TaskRun) {
+func updatePipelineRunStatusFromSlices(ctx context.Context, logger *zap.SugaredLogger, pr *v1beta1.PipelineRun, taskRuns []*v1beta1.TaskRun, runs []*v1alpha1.Run) {
+	taskNameByPipelineTask := make(map[string]string)
```

Comment:
I don't think this variable (`taskNameByPipelineTask`) is necessary

Reply:
It is - we need to ensure we don't give a condition `TaskRun` different names from each of `...FromChildRefs` and `...FromTaskRuns` in the `both` scenario. It's annoying, but it is what it is.

Reply:
got it-- any way to factor out the condition checks and pass them into the `updatePipelineRunStatus` functions? it seems like the `updatePipelineRunStatus` functions could be made more modular: they should just take some info and use it to update the PR status, without having to return data that will be used in other updates

Reply:
Maaaaybe? I think it's just a necessary evil at this point - yet another annoying side effect of `conditions`.

Reply:
maybe a better pattern here could be one function for minimal embedded status, one function for full embedded status, and one function for both embedded status which calls both of the first two. I think this way none of them should need to return anything. thoughts?

Reply:
Hmm...lemme give it a shot and see how it feels.

Reply:
Actually, we'd still have the same problem in that approach - both the minimal and full functions would need to be able to generate names for `condition` `TaskRun`s, so the "both" function would still need to make sure that the name generated in the "minimal" function is still used in the "full" status. 

There's probably some nifty little hack of an approach to deal with this that I'm just not seeing yet. =)

Reply:
The best thing I can come up with so far would be something like when we hit the "oh, we don't already have a `TaskRun` name in `pr.Status.TaskRuns` for this `condition` as part of the `PipelineTask` with this name" in the `...FromTaskRuns` function, we should first iterate over `pr.Status.ChildReferences` to see if there's a child ref there with that `PipelineTaskName`. Would that be preferable to the current approach?

Reply:
hahaha, I just realized that `resources.GetTaskRunName` already does exactly that! Cleaning this up now.

Reply:
And realizing that actually let me clean the logic up even more - see https://github.com/tektoncd/pipeline/pull/4739/commits/c934e216d1e4f151c4e73dde81de9ce458242598

Reply:
(and, for that matter, `kmeta.ChildName` will actually return the same thing every time anyway, so my whole worry about the generated names matching didn't matter in the first place!)